pkg/apis/autoscaling/types.go

```diff
@@ -229,6 +241,20 @@ type MetricSpec struct {
 	// to normal per-pod metrics using the "pods" source.
 	// +optional
 	Resource *ResourceMetricSource
+	// PodResource refers to a resource metric (such as those specified in
+	// requests and limits) known to Kubernetes describing each pod in the
+	// current scale target (e.g. CPU or memory). Such metrics are built in to
+	// Kubernetes, and have special scaling options on top of those available
+	// to normal per-pod metrics using the "pods" source.
+	// +optional
+	PodResource *PodResourceMetricSource `json:"podResource,omitempty"`
+	// ContainerResource refers to a resource metric (such as those specified in
+	// requests and limits) known to Kubernetes describing a single container in each pod of the
+	// current scale target (e.g. CPU or memory). Such metrics are built in to
+	// Kubernetes, and have special scaling options on top of those available
+	// to normal per-pod metrics using the "pods" source.
+	// +optional
+	ContainerResource *ContainerResourceMetricSource `json:"containerResource,omitempty"`
```

Comment:
if we create an HPA instance using this field in 1.19 with this PR merged, it gets serialized to a JSON annotation and persisted in v1 in etcd.

If we update this instance via a 1.18 API server via v2beta2, that API server won't know about this field, and will fail validation.

See https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#adding-unstable-features-to-stable-versions for details on how to add a new field and validation, but feature gate the new field so that we don't allow persisting data that would break an n-1 API server.

For this, we could add the field and validation in 1.19, leave the feature gate off in 1.19, then allow persisting this data in 1.20 by enabling the feature gate.

This means you'll need to enable the feature as part of tests that require persisting the object (integration tests, etc)... see use of SetFeatureGateDuringTest in the codebase for how to do that.